gr-blocks/python/blocks/variable_save_restore.py

```diff
@@ -0,0 +1,143 @@
+            self.variables.append(self.description_variable)
+
+    def set_save_trigger(self, save_trigger):
+        if save_trigger and not self.save_trigger:
+            self.save()
+        self.save_trigger = save_trigger
+
+    def set_restore_trigger(self, restore_trigger):
+        if restore_trigger and not self.restore_trigger:
+            self.restore()
+        self.restore_trigger = restore_trigger
+
+    def handle_msg_save(self, msg):
+        self.save()
+
+    def handle_msg_restore(self, msg):
+        self.restore()
+
+    def get_variable_values(self):
+        """Get values of variables using top-glock getters"""
+        data = {}
+        for name in self.variables:
+            try:
+                data[name] = self._get_variable(self.top_block, name)
+            except AttributeError as e:
+                print("AttributeError:", e)
+                print(f"ERROR: Failed to get variable `{name}`")
+
+        return data
+
+    def _get_variable(self, block, name):
+        if "." in name:
+            subblock, name = name.split(".", maxsplit=1)
+            return self._get_variable(getattr(block, subblock), name)
+        return getattr(block, f"get_{name}")()
+
+    def set_variable_values(self, data, verbose=False):
+        """Set values of variables using top-glock setters"""
+        n = 0
+        for name, val in data.items():
+            if verbose:
+                print(f"{name:20s}    {val}")
+            try:
+                self._set_variable(self.top_block, name, val)
+                n += 1
+            except AttributeError as e:
+                print("AttributeError:", e)
+                print(f"ERROR: Failed to set variable `{name}` to `{val}`")
+        return n
+
+    def _set_variable(self, block, name, val):
+        if "." in name:
+            subblock, name = name.split(".", maxsplit=1)
+            return self._set_variable(getattr(block, subblock), name, val)
+        setter = getattr(block, f"set_{name}")
+        setter(val)
+
+    def save(self):
+        data = self.get_variable_values()
+        self.storage[self.slot] = data
+        try:
+            self.save_storage()
+        except Exception as e:
+            print("Exception:", e)
+            print(f"[VariableSaveRestore] Failed to save variables to slot {self.slot}")
+        else:
+            print(f"[VariableSaveRestore] Saved {len(data)} of {len(self.variables)} variables to slot {self.slot}")
+
+    def restore(self):
+        data = self.storage.get(self.slot, {})
+        data = {k: v for k, v in data.items() if k in self.variables}
+        n = self.set_variable_values(data)
+        print(f"[VariableSaveRestore] Restored {n} of {len(self.variables)} variables from slot {self.slot}")
+
+    def load_storage(self):
+        """Load from disk into memory"""
+        if os.path.exists(self.filename):
+            with open(self.filename, 'rb') as handle:
+                self.storage = pickle.load(handle)
```

Comment:
Since we use YAML elsewhere, we shouldn't use pickle for serialization. This would be the only use.

Reply:
As you like. Changed to yaml safe load

Reply:
Too bad they don't define those tag constants anywhere. For the tuple one, is it important to you that it stays a tuple and not SafeRepresenter's mapping to a list?

Reply:
Yes, save/restore must not unexpectedly change data types.